---
title: "volcano3D Vignette"
author: "Katriona Goldmann"
output: rmarkdown::html_vignette
toc:
  depth: 5
vignette: >
  \VignetteIndexEntry{volcano3D Package}
  \VignetteEngine{knitr::knitr}
  \usepackage[utf8]{inputenc}
---

```{r setup, include = FALSE, echo = FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      fig.align = 'center',
                      dev.args=list(bg="transparent")
)


library(knitr)
library(kableExtra)
source("https://gist.githubusercontent.com/KatrionaGoldmann/3d6dfd6aa4cc5c3940bb72dc49beae02/raw/26e78c78a7a9d096d695c708e8a45830a1d1121a/render_toc.R")
```

<img src="../logo.png" align="right" alt="" width="200" style="border: 0; margin: 5px;" />

# Introduction 

The volcano3D package enables exploration of probes differentially 
expressed between three groups. Its main purpose is for the
visualisation of differentially expressed genes in a three-dimensional
volcano plot. These plots can be converted to interactive visualisations using
[plotly](https://plot.ly/r/). 

This vignette consists of a case study from the PEAC 
rheumatoid arthritis project (Pathobiology of Early Arthritis Cohort). 
The methodology has been published in 
[Lewis, Myles J., et al. 'Molecular portraits of early rheumatoid arthritis 
identify clinical and treatment response phenotypes.' Cell reports 28.9 (2019): 
2455-2470.
(DOI: 10.1016/j.celrep.2019.07.091)](https://doi.org/10.1016/j.celrep.2019.07.091)
with an interactive web tool available at 
[https://peac.hpc.qmul.ac.uk](https://peac.hpc.qmul.ac.uk).  

There are also supplementary vignettes with further information on:

- [creating the input pvalues data frame](https://katrionagoldmann.github.io/volcano3D/articles/pvalues_generator.html)
- [using the volcano3D package to create and deploy a shiny app](https://katrionagoldmann.github.io/volcano3D/articles/shiny_builder.html)

## Getting Started

### Prerequisites

* [ggplot2](https://CRAN.R-project.org/package=ggplot2)
* [ggpubr](https://CRAN.R-project.org/package=ggpubr)
* [ggrepel](https://CRAN.R-project.org/package=ggrepel)
* [plotly](https://CRAN.R-project.org/package=plotly)

### Install from CRAN

[![CRAN status](https://www.r-pkg.org/badges/version/volcano3D)](https://cran.r-project.org/package=volcano3D)

```{r, eval = FALSE}
install.packages("volcano3D")
```

### Install from Github

[![GitHub tag](https://img.shields.io/github/tag/KatrionaGoldmann/volcano3D.svg)](https://GitHub.com/KatrionaGoldmann/volcano3D/tags/)

```{r, eval = FALSE}
library(devtools)
install_github("KatrionaGoldmann/volcano3D")
```

### Load the package

```{r}
library(volcano3D)
```

```{r, echo=FALSE}
library(ggpubr)
library(plotly)
```


---

# Dictionary 

Variables used in this vignette:


| Variable | Definition   |
|-----------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <b>contrast</b> | the variable by which samples can be split into three groups. |
| <b>groups</b> | the three levels/categories of the contrast variable. These should not contain underscores.|
| <b>comparison</b> | two groups between which a statistical test can be performed. There should be three comparisons total. For the examples outlined in this vignette we look at comparisons: 'lymphoid-myeloid', 'lymphoid-fibroid' and 'myeloid-fibroid'. |
| <b>p</b> | p value |
| <b>FC</b> | fold change |
| <b>padj</b> | adjusted p value |
| <b>suffix</b> | the tail word in a column name. In this package it states the statistical parameter (e.g. _logFC is the log FC variable).  |
| <b>prefix</b> | the leading word in a column name. In this package it states the statistical test (e.g. LRT is the likelihood ratio test). |
| <b>polar</b> | A polar coordinates object, of S4 class, containing the expression data, sample data, pvalues and polar coordinates. |



# Example 1. Synovial Gene Data

This vignette will demonstrate the power of this package using a basic example
from the [PEAC data set](http://www.peac-mrc.mds.qmul.ac.uk). Here we will focus
on the synovial data from this cohort.

Using the synovial biopsies from PEAC we can create a polar object for
differentially expressed genes. The sample data used in this vignette can be 
loaded from the [volcano3Ddata package](https://github.com/KatrionaGoldmann/volcano3Ddata). 


```{r, eval=FALSE}
devtools::install_github("KatrionaGoldmann/volcano3Ddata")
library(volcano3Ddata)
data("syn_data")
```

```{r, echo=FALSE}
library(volcano3Ddata)
data("syn_data")
```

Samples in this cohort fall into three pathotype groups: 

```{r}
kable(table(syn_metadata$Pathotype), col.names = c("Pathotype", "Count"))
```

In this example we are interested in genes that are differentially expressed 
between each of these groups. 

First we will set up a polar object, using the polar_coords function, which
uses:

```{r, echo=FALSE}
mytable = data.frame(
  sampledata = c("sampledata\ 
  \n\n(required)", 
  "This shows information for each sample in rows and must contain:\ 
                \n * an ID column: Containing the sample IDs. This must be titled ‘ID’.\ 
                \n * a contrast column: Indicates which of the three groups each sample belongs to.\ 
                \n \n"), 
  contrast = c("contrast\ 
               \n\n(required)", 
               "The column name in sampledata which contains the three-level factor used for contrast"), 
  pvalues = c("pvalues\ 
              \n\n(required)", 
              "the pvalues data.frame which contains the statistical\
                significance of probes between groups. This contains: \
              \n * three pvalue columns: one for each comparison with \
              column names of format `paste(groups[i], groups[j], p_col_suffix, sep='_')`.\ 
              We recommend \
              using 'limma' or 'DESeq' pipelines to calculate these pvalues for \
              gene expression.\ 
              \n * _optional_ fold change columns: one for each comparison with\
              column names of format `paste(groups[i], groups[j], fc_col_suffix, sep='_')`  \
              \n * _optional_ adjusted pvalue columns: one for each comparison\
              with column names of format `paste(groups[i], groups[j], padj_col_suffix, sep='_')` \
              \n * an _optional_ multi-group pvalue column: from a multi-group\
              test with column name of the form `paste0(multi_group_prefix, '_', p_col_suffix)`.This is typically\
              generated using ANOVA or likelihood ratio tests between all three groups. \
              \n * an _optional_ multi-group adjusted pvalue column: from a\
              multi-group test (column names of form `paste0(multiGroupPrefix, '_', padjColSuffix)`). "),
  exp = c("expression\ 
              \n\n(required)", 
          "A data frame or matrix containing the expression data. This \
           is used to calculate z-score and fold change, therefore it should be a \
           normalised expression object such as log transformed or variance stabilised."),
  groups = c("groups", "The groups to be compared (in order). If NULL this \
             defaults to `levels(sampledata[, 'contrasts'])`. These must not contain underscores."), 
  pcsuff = c("p_col_suffix", "The suffix of column names with pvalues (default is 'pvalue'). This must not contain underscores."),
  padjsuff = c("padj_col_suffix", "The suffix of column names with adjusted pvalues (default\
  is 'padj').  This must not contain underscores. If NULL the adjusted pvalue is calculated using `p_col_suffix` and\
  `pvalue_method`."),
  padjmeth = c("padjust_method", "The method to calculate adjusted pvalues if not already\
  provided. Must be one of c('holm', 'hochberg', 'hommel', 'bonferroni', 'BH',\
  'BY', 'fdr', 'none'). Default is 'BH'."),
  gc_suff = c("fc_col_suffix", "The suffix of column names with log(fold change) values\
  (default is 'logFC'). This must not contain underscores."),
  mpref = c("multi_group_prefix", "The prefix for columns containing statistics for a\
  multi-group test (this is typically a likelihood ratio test or ANOVA). Default\
  is NULL. This must not contain underscores."),
  lab = c("label_column", "A column name in pvalues which is to be used to label markers\
  of interest at plotting stage. If NULL the rownames will be used.")
)

kable(t(mytable), row.names = FALSE, col.names = c("Variable", "Details")) %>%
  kable_styling(font_size=11)
```

We can map the PEAC data to polar coordinates using: 

```{r}
syn_polar <- polar_coords(sampledata = syn_metadata,
                          contrast = "Pathotype",
                          pvalues = syn_pvalues,
                          expression = syn_rld,
                          p_col_suffix = "pvalue",
                          padj_col_suffix = "padj",
                          fc_col_suffix = "log2FoldChange",
                          multi_group_prefix = "LRT",
                          non_sig_name = "Not Significant",
                          significance_cutoff = 0.01,
                          label_column = NULL,
                          fc_cutoff = 0.1, 
                          cutoff_criteria="pvalue")
```

This creates a polar class object with slots for: sampledata, contrast, pvalues, 
multi_group_test, expression, polar and non_sig_name. The pvalues slot which 
should have a data frame with statistics for each comparison - pvalue and 
optional logarithmic fold change and adjusted pvalue statistics. In this case 
we are including fold change for downstream visualisations: 

```{r, eval=FALSE}
head(syn_polar@pvalues)
```

```{r, echo=FALSE}
head(syn_polar@pvalues) %>%
  kable(col.names = gsub("_l", " _l", gsub("_p", " _p", colnames(syn_polar@pvalues))) )%>%
  kable_styling(font_size=6)
```

The `sig` column in `syn_polar@polar` allows us to determine relative
differences in expression between groups (in this case pathotypes). The '+'
indicates which pathotypes are significantly 'up' compared to others. For
example:

* genes labelled 'Lymphoid+' are significantly up in Lymphoid vs
Myeloid **and** Lymphoid vs Fibroid.

* genes up in two pathotypes such as 'Lymphoid+Myeloid+' are up in both
Lymphoid **and** Myeloid, therefore Lymphoid vs
Fibroid and Myeloid vs Fibroid are statistically significant.

* genes which show no significant difference between pathotypes are classed
according to `non_sig_name`

This gives us:

```{r, eval=FALSE}
setNames(data.frame(table(syn_polar@polar$sig)), c("Significance", "Frequency"))
```

```{r, echo=FALSE}
table(syn_polar@polar$sig) %>%
  kable(col.names = c("Significance", "Frequency")) %>%
  kable_styling(full_width = F)
```

To subset and view objects of specific significance groups you can use the 
`significance_subset` function. 

```{r}
pvalues_subset <- significance_subset(syn_polar, 
                                      significance = c("Lymphoid+", "Myeloid+"), 
                                      output="pvalues")

polar_subset <- significance_subset(syn_polar, 
                                    significance = c("Lymphoid+", "Myeloid+"), 
                                    output="polar")

head(pvalues_subset) %>%
  kable() %>%
  kable_styling(full_width = F)
```


## Volcano Plots

If there is a fold change column previously provided, we can now investigate
the comparisons between pathotypes using the
volcano_trio function. This creates three
[ggplot](https://ggplot2.tidyverse.org/reference/ggplot.html) outputs.

```{r, fig.height=2.8, fig.width=10}
syn_plots <- volcano_trio(polar = syn_polar,
                          sig_names = rep(c("not significant","significant"), each=2),
                          colours = rep(c( "slateblue1", "grey60"), each=2),
                          colour_scheme="none",
                          text_size = 9,
                          marker_size=1.5,
                          shared_legend_size = 0.9,
                          label_rows = c("SLAMF6", "PARP16", "ITM2C"),
                          fc_line = FALSE,
                          share_axes = FALSE)

syn_plots$All
```

Alternatively using the polar significance levels

```{r, fig.height=2.8, fig.width=10}
syn_plots <- volcano_trio(polar = syn_polar,
                          colour_scheme="polar",
                          colours = c('green3', 'cyan',  'blue', 
                                      'purple', 'red', 'gold2', 'grey60'),
                          text_size = 9,
                          marker_size=1.5,
                          shared_legend_size = 0.9,
                          label_rows = c("SLAMF6", "PARP16", "ITM2C"),
                          fc_line = FALSE,
                          p_line = FALSE,
                          share_axes = FALSE)

syn_plots$All
```

or the upregulated group within the standard volcano plot: 

```{r, fig.height=2.8, fig.width=10}
syn_plots <- volcano_trio(polar = syn_polar,
                          colour_scheme="upregulated",
                          colours = c('green3', 'cyan',  'blue', 
                                      'purple', 'red', 'gold2', 'grey60'),
                          text_size = 9,
                          marker_size=1.5,
                          shared_legend_size = 0.9,
                          label_rows = c("SLAMF6", "PARP16", "ITM2C"),
                          fc_line = FALSE,
                          p_line = FALSE,
                          share_axes = FALSE)

syn_plots$All
```

These use the pvalue column to colour-code significance. Alternatively we can use 
the adjusted pvalues (right): 

```{r, fig.height=2.8, fig.width=6.6}
p <- volcano_trio(polar = syn_polar,
                  colour_scheme="upregulated",
                  colours = c('green3', 'cyan',  'blue', 
                              'purple', 'red', 'gold2', 'grey60'),
                  text_size = 9,
                  marker_size=2,
                  shared_legend_size = 0.9,
                  label_rows = c("SLAMF6", "PARP16", "ITM2C"),
                  fc_line = FALSE,
                  p_line = TRUE,
                  share_axes = FALSE, 
                  marker_outline_colour="black") 

padj <- volcano_trio(polar = syn_polar,
                     colour_scheme="upregulated",
                     colours = c('green3', 'cyan',  'blue', 
                                 'purple', 'red', 'gold2', 'grey60'),
                     text_size = 9,
                     marker_size=2,
                     shared_legend_size = 0.9,
                     label_rows = c("SLAMF6", "PARP16", "ITM2C"),
                     fc_line = FALSE,
                     p_line = TRUE,
                     share_axes = FALSE, 
                     marker_outline_colour="black", 
                     cutoff_criteria = "padj") 

ggarrange(p$Fibroid_Lymphoid + labs(subtitle="Pvalue cutoff") + lims(x=c(-10, 10)), 
          padj$Fibroid_Lymphoid + labs(subtitle="Adjusted pvalue cutoff") + lims(x=c(-10, 10)), 
          common.legend = T)
```


## Radial Plots

The differential expression can now be visualised on an interactive radar plot 
using radial_plotly. 

Using the `hover` parameter (implemented v1.0.2) the plotly hover information 
can be altered by referencing the columns available in polar@polar and 
polar@pvlaues i.e.: 

```{r}
unique(c(colnames(syn_polar@polar), colnames(syn_polar@pvalues)))
```

For example to show all pvalues on hover we can use: 

```{r}
hovertext = "paste(label, 
              '\nL vs M pvalue:', format(Lymphoid_Myeloid_pvalue, digits=3), 
              '\nM vs F pvalue:', format(Myeloid_Fibroid_pvalue, digits=3), 
              '\nL vs F pvalue:', format(Fibroid_Lymphoid_pvalue, digits=3))"
```

Similarly the `labelRows` variable allows any markers of interest to be 
labelled:

```{r, out.height="100%", out.width="100%"}
radial_plotly(polar = syn_polar, 
              label_rows = c("SLAMF6", "PARP16", "ITM2C"), 
              hover=hovertext)
```

By hovering over certain points you can also determine genes for
future interrogation.

Similarly we can create a static ggplot image using radial_ggplot:

```{r, fig.height=4.5, fig.width=7}
radial_ggplot(polar = syn_polar,
              label_rows = c("SLAMF6", "FMOD"),
              marker_size = 2.3,
              legend_size = 10) +
  theme(legend.position = "right")
```

Alternatively a continuous colour scale can be produced by converting the angle
to a hsv variable. This angle can be offset by `continuous_shift`.

```{r, fig.height=4.5, fig.width=7}
radial_ggplot(polar = syn_polar,
              label_rows = c("SLAMF6", "PARP16", "ITM2C"),
              marker_size = 2.3,
              marker_alpha=0.7,
              colour_scale = "continuous",
              continuous_shift=1.33,
              legend_size = 10) 
```


## Boxplots

We can then interrogate any one specific variable as a boxplot, to investigate
these differences. This is build using either ggplot2 or plotly so can easily be edited by the
user to add features. Using plotly:

```{r, fig.height = 3.7, fig.width=9}
plot1 <- boxplot_trio(syn_polar,
                      value = "FAM92B",
                      text_size = 7,
                      test = "polar_padj",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      box_colours = c("blue", "red", "green3"),
                      step_increase = 0.2,
                      plot_method='plotly')

plot2 <- boxplot_trio(syn_polar,
                      value = "SLAMF6",
                      text_size = 7,
                      test = "polar_multi_pvalue",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      box_colours = c("blue", "red", "green3"), 
                      plot_method='plotly') 

plot3 <- boxplot_trio(syn_polar,
                      value = "PARP16",
                      text_size = 7,
                      stat_size=2.5,
                      test = "t.test",
                      levels_order = c("Myeloid", "Fibroid"),
                      box_colours = c("pink", "gold"), 
                      plot_method='plotly') 

plotly::subplot(plot1, plot2, plot3, titleY=TRUE, margin=0.05)
```

Or using ggplot

```{r, fig.height = 3.2, fig.width=7}
plot1 <- boxplot_trio(syn_polar,
                      value = "FAM92B",
                      text_size = 7,
                      test = "polar_pvalue",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      box_colours = c("blue", "red", "green3"),
                      step_increase = 0.1)

plot2 <- boxplot_trio(syn_polar,
                      value = "SLAMF6",
                      text_size = 7,
                      test = "polar_multi_pvalue",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      box_colours = c("blue", "red", "green3")) 

plot3 <- boxplot_trio(syn_polar,
                      value = "PARP16",
                      text_size = 7,
                      stat_size=2.5,
                      test = "t.test",
                      levels_order = c("Myeloid", "Fibroid"),
                      box_colours = c("pink", "gold")) 

ggarrange(plot1, plot2, plot3, ncol=3)
```

## Three Dimensional Volcano Plots

The final thing we can look at is the 3D volcano plot which projects
differential gene expression onto cylindrical coordinates.

```{r, fig.width=7}
p <- volcano3D(syn_polar,
               label_rows = c("SLAMF6", "PARP16", "ITM2C"),
               label_size = 10,
               radial_axis_title_offset = 1.1,
               z_axis_title_offset = 1.2,
               colour_code_labels = F,
               label_colour = "black",
               xy_aspectratio = 1.2,
               z_aspectratio = 0.9, 
               axes_from_origin=FALSE,  
               plot_height = 500)

p
```


We can alter the colour code using the colours parameter. These are assigned in 
order group1+, group1+group2+, group2+, group2+group3+, group3+, group1+group3+.

```{r, fig.width=7}
volcano3D(syn_polar,
               label_rows = c("SLAMF6", "PARP16", "ITM2C"),
               label_size = 10,
               xy_aspectratio = 1.2,
               z_aspectratio = 0.9, 
               colours = c("grey60", "grey60", "blue", 
                           "grey60", "grey60", "grey60"), 
               hover_text = "paste0(Name, '\n', sig)",
               plot_height=500)
```

## Volcano4D: Rotating plots

In order to rotate/spin your 3D volcano plot you can use the `volcano4D` function. This adds a button to the plotly modeBar (hover top right) which allows you to rotate the plot. 

```{r, fig.width=7}
volcano4D(p)
```


# Saving Plotly Plots

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">


## Static Images

There are a few ways to save plotly plots as static images. Firstly plotly 
offers a download button ( <i class="fa fa-camera" aria-hidden="true"></i> ) in the figure mode bar (appears top right). 
By default this saves images as png, however it is possible to convert to svg, jpeg or webp using:

```{r, eval=FALSE}
p %>% plotly::config(toImageButtonOptions = list(format = "svg"))
```

Alternatively, if orca command-line utility is installed, this can also be used to
save plotly images. To install follow the instructions
[here](https://github.com/plotly/orca#installation).

```{r, eval = FALSE}
orca(p, "./volcano_3d_synovium.svg", format = "svg")
```

## Interactive HTML

The full plotly objects can be saved to HTML by converting them to widgets and 
saving with the htmlwidgets package: 

```{r, eval=FALSE}
htmlwidgets::saveWidget(as_widget(p), "volcano3D.html")
```


# Altering the Plots

## Altering the Grid

By default volcano3D generates a grid using 12 spokes. If you wish to override
any of these variables it is possible to pass in your own grid. This is also
possible for the 3D radial plots where the z elements can be left as NULL.

By manually creating a grid object it is possible to change the tick points on
axes as well as the number of radial spokes (n_spokes). The default ticks are calculated from r_vector and z_vector using the
`pretty` function. This can be overwritten by passing tick points in
r_axis_ticks and z_axis_ticks.

For example, we can decrease the number of radial spokes to 4, while altering the z axis ticks:

```{r}
four_grid = polar_grid(r_vector=syn_polar@polar$r_zscore,
                       z_vector=NULL,
                       r_axis_ticks = NULL,
                       axes_from_origin = FALSE, 
                       z_axis_ticks = c(0, 8, 16, 32),
                       n_spokes = 4)
```

We can inspect the grid using `show_grid()` which creates both the polar and 
cylindrical coordinate system:

```{r, fig.width=7}
p <- show_grid(four_grid)

p$cylindrical
```

and pass it into the plotting functions:

```{r, fig.width=7, eval=FALSE}
volcano3D(syn_polar,
          grid = four_grid,
          label_rows = c("SLAMF6", "PARP16", "ITM2C"),
          label_size = 10,
          xy_aspectratio = 1,
          z_aspectratio = 0.9)
```

For example to extend the radial axis and increase the number of spokes in 2D
we can apply:

```{r, fig.height=5, fig.widht=5}
new_grid = polar_grid(r_vector=NULL,
                      z_vector=-log10(syn_polar@pvalues$LRT_pvalue),
                      r_axis_ticks = c(1, 2, 3),
                      z_axis_ticks = NULL,
                      n_spokes = 24)
```

## Altering the Annotations

To amend or reposition the plotly labels it is possible to use the editable parameter. 
This allows you to move annotation features (labels), change text including 
legends and titles. Click on a feature to change it:

```{r, fig.height=4.5, fig.width=7, eval=FALSE}
p = radial_plotly(polar = syn_polar, 
                  label_rows = c("SLAMF6", "PARP16", "ITM2C"))

p %>% config(editable = TRUE)
```

## Altering the Labels

By default the rownames of the pvalues object is used to label markers of
interest. This can be altered by amending the label column in the polar object:

```{r, fig.height=4.5, fig.width=7}
syn_polar@polar$label <- paste0(syn_polar@polar$label, "!")

radial_ggplot(polar = syn_polar,
              label_rows = c("SLAMF6", "PARP16", "ITM2C"),
              grid = new_grid,
              marker_size = 2.3,
              legend_size = 10) +
  theme(legend.position = "right")
```

# Example 2. Synovial Modular Data

We can collapse this example into a modular analysis using a list of gene sets.
In this example we have used the blood transcript modules curated by Li et. al.
in ['Li, S., Rouphael, N., Duraisingham, S., Romero-Steiner, S., Presnell, S.,
Davis, C., ... & Kasturi, S. (2014). Molecular signatures of antibody responses
derived from a systems biology study of five human vaccines. Nature immunology,
15(2), 195.'](https://www.nature.com/ni/journal/v15/n2/abs/ni.2789.html). The
pvalues were generated using
[QuSAGE methodology](http://clip.med.yale.edu/qusage/).

## Creating polar coordinates

The modular analysis can be loaded through Li_pvalues:

```{r}
data(Li_pvalues)

syn_mod_polar <- polar_coords(sampledata = syn_metadata,
                              contrast = "Pathotype",
                              pvalues = syn_mod_pvalues,
                              p_col_suffix = "p.value",
                              padj_col_suffix = "q.value",
                              fc_col_suffix = "logFC",
                              multi_group_prefix = NULL,
                              expression = syn_mod,
                              significance_cutoff = 0.01,
                              fc_cutoff = 0.1,
                              cutoff_criteria = "pvalue"
)
```


## Volcano Plots

```{r, message = FALSE, fig.height=2.8, fig.width=10}
syn_mod_plots <- volcano_trio(polar = syn_mod_polar,
                              label_rows = c("M156.0", "M37.2"),
                              shared_legend_size = 1,
                              sig_names = c("Not Sig",
                                            paste("Padj <", 0.05),
                                            paste("|FC| >", 1),
                                            paste("Padj <", 0.05,
                                                  "&\n|FC| >", 1)),
                              share_axes = FALSE)

syn_mod_plots$All
```

## Radial Plots


```{r, out.height="100%", out.width="90%", fig.height=8, eval=FALSE, fig.width=10}
radial_plotly(polar = syn_mod_polar,
              axes_from_origin = FALSE, 
              label_rows = c("M156.0", "M37.2"))
```

Or a ggplot static image using radial_ggplot:

```{r, warning = FALSE, fig.height=4.5, fig.width=7}
radial_ggplot(polar = syn_mod_polar,
              label_rows = c("M156.0", "M37.2"),
              marker_size = 2.7,
              label_size = 5,
              axis_lab_size = 3,
              axis_title_size = 5,
              legend_size = 10) 
```

## Boxplots

We can then interrogate specific modules with boxplot_trio:

```{r, fig.height = 2.8, fig.width=5}
plot1 <- boxplot_trio(syn_mod_polar,
                      value = "M156.0",
                      test = "wilcox.test",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      box_colours = c("blue", "red", "green3"))

plot2 <- boxplot_trio(syn_mod_polar,
                      value = "M37.2",
                      test = "wilcox.test",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      box_colours = c("blue", "red", "green3"))

ggpubr::ggarrange(plot1, plot2)
```

---

# Citation

If you use this package please cite as:

```{r}
citation("volcano3D")
```

or using:

> Lewis, Myles J., et al. 'Molecular portraits of early rheumatoid arthritis
identify clinical and treatment response phenotypes.' Cell reports 28.9 (2019):
2455-2470.


