<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>volcano3D: 2x3-way analysis • volcano3D</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="volcano3D: 2x3-way analysis">
<meta property="og:description" content="volcano3D">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-158679490-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158679490-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">volcano3D</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Extended_Vignette.html">Volcano3D Vignette</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Additional Tutorials</li>
    <li>
      <a href="../articles/Vignette_2x3.html">volcano3D: 2x3-way analysis</a>
    </li>
    <li>
      <a href="../articles/shiny_builder.html">Using Volcano3D to create Shiny Apps</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">
    <span class="fas fa-newspaper"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="far fa-file-code"></span>
     
    Reference Manual
  </a>
</li>
<li>
  <a href="https://github.com/KatrionaGoldmann/volcano3D">
    <span class="fab fa-github"></span>
     
    github
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Vignette_2x3_files/header-attrs-2.21/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>volcano3D: 2x3-way analysis</h1>
                        <h4 class="author">Myles Lewis, Katriona
Goldmann, Elisabetta Sciacca</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KatrionaGoldmann/volcano3D/blob/master/vignettes/Vignette_2x3.Rmd"><code>vignettes/Vignette_2x3.Rmd</code></a></small>
      <div class="hidden name"><code>Vignette_2x3.Rmd</code></div>

    </div>

    
    
<style>
  .title{
    display: none;
  }
</style>
<div id="x3-way-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#x3-way-analysis" class="anchor"></a>2x3-way analysis</h2>
<p><img src="https://katrionagoldmann.github.io/volcano3D/logo.png" align="right" alt="" width="200" hspace="50" vspace="50" style="border: 0;"></p>
<p>The main work flow for using volcano3D is ideal for comparing high
dimensional data such as gene expression or other biological data across
3 classes, and this is covered in the main vignette. However, an
alternative use of 3-way radial plots and 3d volcano plots is for
2x3-way analysis. In this type of analysis there is a binary factor such
as drug response (responders vs non-responders) and a 2nd factor with 3
classes such as a trial with 3 drugs.</p>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>This vignette shows analysis from the STRAP trial in rheumatoid
arthritis, in which patients were randomised to one of three drugs
(etanercept, rituximab, tocilizumab). Clinical response to treatment (a
binary outcome) was measured after 16 weeks. Gene expression data from
RNA-Sequencing (RNA-Seq) of synovial biopsies from the patients’
inflamed joints was performed at baseline. RNA-Seq data is count based
and overdispersed, and thus is typically best modelled by a negative
binomial distribution.</p>
</div>
<div id="deseq2-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#deseq2-pipeline" class="anchor"></a>DESeq2 pipeline</h2>
<p>Differential gene expression to compare the synovial gene expression
between responders vs non-responders is performed using the Bioconductor
package DESeq2. Since there are 3 distinct drugs this becomes a
2x3-factor analysis.</p>
<p>The following code shows set up and 2x3-way analysis in DESeq2
followed by generation of a ‘volc3d’ class results object for plotting
and visualisation of the results.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://katrionagoldmann.github.io/volcano3D/index.html">volcano3D</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Basic DESeq2 set up</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html">rnbinom</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">3000</span>, mu<span class="op">=</span><span class="fl">100</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">0.5</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">cond</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, each<span class="op">=</span><span class="fl">5</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="st">'B'</span>, <span class="st">'C'</span><span class="op">)</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">resp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, each<span class="op">=</span><span class="fl">15</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'non.responder'</span>, <span class="st">'responder'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>drug <span class="op">=</span> <span class="va">cond</span>, response <span class="op">=</span> <span class="va">resp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Full dataset object construction</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromMatrix</a></span><span class="op">(</span><span class="va">counts</span>, <span class="va">metadata</span>, <span class="op">~</span><span class="va">response</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform 3x DESeq2 analyses comparing binary response for each drug</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseq_2x3.html">deseq_2x3</a></span><span class="op">(</span><span class="va">dds</span>, <span class="op">~</span><span class="va">response</span>, <span class="st">"drug"</span><span class="op">)</span></span></code></pre></div>
<p>The design formula can contain covariates, however the main contrast
should be the last term of the formula as is standard in DESeq2
analysis.</p>
<p>The function <code><a href="../reference/deseq_2x3.html">deseq_2x3()</a></code> returns a list of 3 DESeq2
objects containing the response analysis for each of the three drugs.
These response vs non-response differential expression comparisons can
be quickly visualised with the <code><a href="https://rdrr.io/pkg/easylabel/man/easyVolcano.html">easyVolcano()</a></code> function from
the R package <code>easylabel</code>, which is designed for DESeq2 and
limma objects and uses an interactive R/shiny interface.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">easylabel</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">res</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>  <span class="co"># results for the first drug  </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/easylabel/man/easyVolcano.html">easyVolcano</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<p>The <code>deseq_2x3</code> output is passed to
<code><a href="../reference/deseq_2x3_polar.html">deseq_2x3_polar()</a></code> to generate a <code>volc3d</code> class
object for plotting. Thus the 3-way radial plot and 3d volcano plot
simplify visualisation of the 2x3-way analysis, replacing 3 volcano
plots with a single radial plot or 3d volcano plot.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate polar object</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseq_2x3_polar.html">deseq_2x3_polar</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2d plot</span></span>
<span><span class="fu"><a href="../reference/radial_plotly.html">radial_plotly</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3d plot</span></span>
<span><span class="fu"><a href="../reference/volcano3D.html">volcano3D</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="real-world-example" class="section level2">
<h2 class="hasAnchor">
<a href="#real-world-example" class="anchor"></a>Real world example</h2>
<p>The example below using real data from the STRAP trial demonstrates a
3-way radial plot highlighting the relationship between genes
significantly increased in responders across each of the three drugs.
The pipe to <code><a href="https://rdrr.io/pkg/plotly/man/toWebGL.html">toWebGL()</a></code> is used to speed up plotting by using
webGL instead of SVG in plotly scatter plots. Alternatively, a ggplot2
version can be plotted using <code><a href="../reference/radial_ggplot.html">radial_ggplot()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseq_2x3_polar.html">deseq_2x3_polar</a></span><span class="op">(</span><span class="va">data1</span><span class="op">)</span></span>
<span><span class="va">labs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'TNXA'</span>, <span class="st">'FLG2'</span>, <span class="st">'MYBPC1'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/radial_plotly.html">radial_plotly</a></span><span class="op">(</span><span class="va">obj</span>, type<span class="op">=</span><span class="fl">2</span>, label_rows <span class="op">=</span> <span class="va">labs</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plotly/man/toWebGL.html">toWebGL</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_2x3_pos.png" width="70%" style="border: 0;"></p>
<p>A 3d volcano plot can be generated too.</p>
<p><img src="volc3d_2x3.png" width="70%" style="border: 0;"></p>
<p>To show genes which are significantly increased in non-responders for
each drug, we reprocess the object using the argument
<code>process = "negative"</code> as this leads to a different colour
scheme for each gene. The polar coordinates are essentially flipped as
each axis represents increased expression in non-responders.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseq_2x3_polar.html">deseq_2x3_polar</a></span><span class="op">(</span><span class="va">data1</span>, process <span class="op">=</span> <span class="st">"negative"</span><span class="op">)</span></span>
<span><span class="va">labs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'MS4A1'</span>, <span class="st">'TNXA'</span>, <span class="st">'FLG2'</span>, <span class="st">'MYBPC1'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/radial_plotly.html">radial_plotly</a></span><span class="op">(</span><span class="va">obj</span>, type<span class="op">=</span><span class="fl">2</span>, label_rows <span class="op">=</span> <span class="va">labs</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plotly/man/toWebGL.html">toWebGL</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_2x3_neg.png" width="70%" style="border: 0;"></p>
</div>
<div id="custom-2x3-way-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#custom-2x3-way-analysis" class="anchor"></a>Custom 2x3-way analysis</h2>
<p>The above workflow is designed for RNA-Seq count data. For other raw
data types, the function <code><a href="../reference/polar_coords_2x3.html">polar_coords_2x3()</a></code> is used to map
attributes to polar coordinates in a 2x3-way analysis.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">polar_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/polar_coords_2x3.html">polar_coords_2x3</a></span><span class="op">(</span><span class="va">vstdata</span>, <span class="va">metadata</span>, <span class="st">"ACR.response.status"</span>,</span>
<span>                              <span class="st">"Randomised.Medication"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/radial_plotly.html">radial_plotly</a></span><span class="op">(</span><span class="va">polar_obj</span>, type<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/volcano3D.html">volcano3D</a></span><span class="op">(</span><span class="va">polar_obj</span><span class="op">)</span></span></code></pre></div>
<p>In the above code example <code>vstdata</code> contains transformed
(approximately Gaussian) gene expression data with genes in columns and
samples in rows. <code>metadata</code> is a dataframe of sample
information with samples in rows. <code>"ACR.response.status"</code>
refers the the binary response column in <code>metadata</code>.
<code>"Randomised.Medication"</code> refers to the 3-way group column in
<code>metadata</code>.</p>
<p><code><a href="../reference/polar_coords_2x3.html">polar_coords_2x3()</a></code> accepts raw data and performs all the
calculations needed to generate coordinates, colours etc for plotting
radial 3-way plots or 3d volcano plots. This differs from the original
<code><a href="../reference/polar_coords.html">polar_coords()</a></code> function in that it is the mean difference
between the 2-way response outcome which is used to map coordinates
along each axis for each of the 3 groups for unscaled polar
coordinates.</p>
<p>Scaled polar coordinates are generated using the <em>t</em>-statistic
for each group comparison.</p>
<p>There is no straightforward equivalent of a one-way ANOVA or
likelihood ratio test to use for the <em>z</em> axis as is used in the
simpler 3-way analysis for standard 3-way radial plots/ 3d volcano
plots. So instead the <em>z</em> axis in a 2x3-way analysis uses the
smallest p-value from each of the 3 paired comparisons. This p-value is
transformed as usual as -log<sub>10</sub>(p).</p>
<p>A table of p-values can be supplied by the user. But if a table of
p-values is absent, p-values are automatically calculated by
<code><a href="../reference/polar_coords_2x3.html">polar_coords_2x3()</a></code>. Each of the 3 groups is subsetted and
pairwise comparisons of the binary response factor are performed.
Options here include unpaired <em>t</em>-test and Wilcoxon test (see
<code><a href="../reference/calc_stats_2x3.html">calc_stats_2x3()</a></code>).</p>
<p>The colour scheme is not as straightforward as for the standard polar
plot and volcano3D plot since genes (or attributes) can be significantly
up or downregulated in the response comparison for each of the 3 groups.
<code>process = "positive"</code> means that genes are labelled with
colours if a gene is significantly upregulated in the response for that
group. This uses the primary colours (RGB) so that if a gene is
upregulated in both red and blue group it becomes purple etc with
secondary colours. If the gene is upregulated in all 3 groups it is
labelled black. Non-significant genes are in grey.</p>
<p>With <code>process = "negative"</code> genes are coloured when they
are significantly downregulated. With <code>process = "two.sided"</code>
the colour scheme means that both significantly up- and down-regulated
genes are coloured with down-regulated genes labelled with inverted
colours (i.e. cyan is the inverse of red etc). However, significant
upregulation in a group takes precedence.</p>
<p>With <code>process = "negative"</code> the polar coordinates are also
flipped as each axis now represents upregulated expression in
non-responders.</p>
<p>The end result is a <code>volc3d</code> class object for downstream
plotting.</p>
</div>
<div id="forest-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#forest-plots" class="anchor"></a>Forest plots</h2>
<p>Specific genes/variables can be interrogated using a forest plot to
investigate differences in binary response between the 3 groups.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">forest_ggplot</span>(obj, <span class="fu">c</span>(<span class="st">"MS4A1"</span>, <span class="st">"FLG2"</span>, <span class="st">"SFN"</span>)</span></code></pre></div>
<p><img src="forest.png" width="70%" style="border: 0;"></p>
<p>The related functions <code><a href="../reference/forest_plot.html">forest_plot()</a></code> and
<code><a href="../reference/forest_plot.html">forest_plotly()</a></code> can be used to generate plots in base
graphics or plotly.</p>
<hr>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>volcano3D was developed by the bioinformatics team at the <a href="https://www.qmul.ac.uk/whri/emr/">Experimental Medicine &amp;
Rheumatology department</a> and <a href="https://www.qmul.ac.uk/c4tb/">Centre for Translational
Bioinformatics</a> at Queen Mary University London.</p>
<p>If you use this package please cite as:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"volcano3D"</span><span class="op">)</span></span></code></pre></div>
<pre><code>## 
## To cite package 'volcano3D' in publications use:
## 
##   Goldmann K, Lewis M (2020). _volcano3D: 3D Volcano Plots and Polar
##   Plots for Three-Class Data_.
##   https://katrionagoldmann.github.io/volcano3D/index.html,
##   https://github.com/KatrionaGoldmann/volcano3D.
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {volcano3D: 3D Volcano Plots and Polar Plots for Three-Class Data},
##     author = {Katriona Goldmann and Myles Lewis},
##     year = {2020},
##     note = {https://katrionagoldmann.github.io/volcano3D/index.html, https://github.com/KatrionaGoldmann/volcano3D},
##   }</code></pre>
<p>or using:</p>
<blockquote>
<p>Lewis, Myles J., et al. <em>Molecular portraits of early rheumatoid
arthritis identify clinical and treatment response phenotypes</em>. Cell
reports 28.9 (2019): 2455-2470.</p>
</blockquote>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css">
<style>
  #left {
    float:left;
  }
  #right {
    width:100%;
    position:relative;
    text-align:right;
    align:right;
    float:right;
  }
  *.icon-grey {color: lightgrey}
  j {
    text-align:center;
    vertical-align:middle;
    position: relative;
    padding-bottom: 12px;
  }
  .smadge:after{
    content:"EMR";
    position: absolute;
    background: lightgrey;
    height:1.2rem;
    top:1rem;
    right:-1rem;
    width:2.3rem;
    text-align: center;
    line-height: 1rem;;
    font-size: 1rem;
    color: #cc0099;
    border-radius: 10%;
  }
  .bladge:after{
    content:"C4TB";
    position: absolute;
    background: lightgrey;
    height:1.2rem;
    top:1rem;
    right:-1rem;
    width:2.7rem;
    text-align: center;
    line-height: 1rem;;
    font-size: 1rem;
    color: #cc0099;
    border-radius: 10%;
  }
</style>
<div id="left">
Developed by <a href="https://www.turing.ac.uk/people/enrichment-students/katriona-goldmann">Katriona Goldmann</a> and <a href="https://www.qmul.ac.uk/whri/people/academic-staff/items/lewismyles.html">Myles Lewis</a>
</div>

<div id="right">
<a href="https://www.qmul.ac.uk/whri/emr/"><j class="fa fa-university fa-2x  icon-grey smadge" aria-hidden="false" style="color:#cc0099"></j></a>
<span style="margin-left:1rem"></span>

<a href="https://www.qmul.ac.uk/c4tb/"><j class="fa fa-university fa-2x  icon-grey bladge" aria-hidden="false" style="color:#cc0099"></j></a>
<span style="margin-left:1rem"></span>

<a href="https://towardsdatascience.com/interactive-differential-expression-analysis-with-volcano3d-cf6c37b5aa6a"><i class="fab fa-medium-m fa-2x" style="color:#cc0099"></i></a>
<span style="margin-left:1rem"></span>
<a href="mailto:k.goldmann@qmul.ac.uk" class="fas fa-envelope fa-2x" style="color:#cc0099">
</a><span style="margin-left:1rem"></span>

<a href="https://github.com/KatrionaGoldmann"><i class="fab fa-github fa-2x" style="color:#cc0099"></i></a>
<span style="margin-left:1rem"></span>

<a href="https://twitter.com/_katriona"><i class="fab fa-twitter fa-2x" style="color:#cc0099"></i></a>
</div>

      </footer>
</div>

  


  </body>
</html>
