<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="volcano3D">
<title>volcano3D • volcano3D</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="volcano3D">
<meta property="og:description" content="volcano3D">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">volcano3D</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Extended_Vignette.html">volcano3D Vignette</a>
    <a class="dropdown-item" href="../articles/Vignette.html">volcano3D</a>
    <a class="dropdown-item" href="../articles/Vignette_2x3.html">volcano3D: 2x3-way analysis</a>
    <a class="dropdown-item" href="../articles/shiny_builder.html">Building a Shiny App</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KatrionaGoldmann/volcano3D/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="Vignette_files/kePrint-0.0.1/kePrint.js"></script><link href="Vignette_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>volcano3D</h1>
                        <h4 data-toc-skip class="author">Katriona
Goldmann, Myles Lewis</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KatrionaGoldmann/volcano3D/blob/HEAD/vignettes/Vignette.rmd" class="external-link"><code>vignettes/Vignette.rmd</code></a></small>
      <div class="d-none name"><code>Vignette.rmd</code></div>
    </div>

    
    
<style>
  .title{
    display: none;
  }
</style>
<div class="section level2">
<h2 id="volcano3d">volcano3D<a class="anchor" aria-label="anchor" href="#volcano3d"></a>
</h2>
<p><img src="https://katrionagoldmann.github.io/volcano3D/logo.png" align="right" alt="" width="200" hspace="50" vspace="50" style="border: 0;"></p>
<p>The volcano3D package provides a tool for analysis of three-class
high dimensional data. It enables exploration of genes differentially
expressed between three groups. Its main purpose is for the
visualisation of differentially expressed genes in a three-dimensional
volcano plot or three-way polar plot. These plots can be converted to
interactive visualisations using plotly. The 3-way polar plots and 3d
volcano plots can be applied to any data in which multiple attributes
have been measured and their relative levels are being compared across
three classes.</p>
<p>This vignette covers the basic features of the package using a small
example data set. To explore more extensive examples and view the
interactive radial and volcano plots, see the <a href="https://katrionagoldmann.github.io/volcano3D/articles/Extended_Vignette.html" class="external-link">extended
vignette</a> which explores a case study from the PEAC rheumatoid
arthritis trial (Pathobiology of Early Arthritis Cohort). The
methodology has been published in <a href="https://doi.org/10.1016/j.celrep.2019.07.091" class="external-link">Lewis, Myles J., et
al. <em>Molecular portraits of early rheumatoid arthritis identify
clinical and treatment response phenotypes</em>. Cell reports 28.9
(2019): 2455-2470. (DOI: 10.1016/j.celrep.2019.07.091)</a> with an
interactive, searchable web tool available at <a href="https://peac.hpc.qmul.ac.uk" class="external-link">https://peac.hpc.qmul.ac.uk</a>. This
was creating as an <a href="https://shiny.rstudio.com" class="external-link">R Shiny app</a>
and deployed to the web using a server.</p>
<p>There are also supplementary vignettes with further information
on:</p>
<ul>
<li><a href="https://katrionagoldmann.github.io/volcano3D/articles/shiny_builder.html" class="external-link">using
the volcano3D package to create and deploy a shiny app</a></li>
</ul>
<p> <a href="https://lifecycle.r-lib.org/articles/stages.html" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-blue.svg" alt="Lifecycle: Stable"></a> <a href="https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html" class="external-link"><img src="https://img.shields.io/badge/License-GPL%20v2-mediumpurple.svg" alt="License: GPL v2"></a> <a href="https://cran.r-project.org/package=volcano3D" class="external-link"><img src="https://www.r-pkg.org/badges/version/volcano3D" alt="CRAN status"></a> <a href="https://cranlogs.r-pkg.org/badges/grand-total/volcano3D" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/volcano3D?color=orange" alt="Downloads"></a> <a href="https://github.com/KatrionaGoldmann/volcano3D/blob/master/NEWS.md" class="external-link"><img src="https://img.shields.io/badge/last%20commit-2022--09--09-turquoise.svg" alt="2022-09-09"></a> <a href="https://travis-ci.org/KatrionaGoldmann/volcano3D" class="external-link"><img src="https://travis-ci.org/KatrionaGoldmann/volcano3D.svg?branch=master" alt="Build Status"></a> <a href="https://GitHub.com/KatrionaGoldmann/volcano3D/issues/" class="external-link"><img src="https://img.shields.io/github/issues/KatrionaGoldmann/volcano3D.svg" alt="GitHub issues"></a> <a href="https://GitHub.com/KatrionaGoldmann/volcano3D/issues/" class="external-link"><img src="https://img.shields.io/travis/KatrionaGoldmann/volcano3D.svg" alt="build"></a></p>
<div class="section level3">
<h3 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h3>
<div class="section level4">
<h4 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h4>
<ul>
<li><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></li>
<li><a href="https://CRAN.R-project.org/package=ggpubr" class="external-link">ggpubr</a></li>
<li><a href="https://CRAN.R-project.org/package=plotly" class="external-link">plotly</a></li>
</ul>
</div>
<div class="section level4">
<h4 id="install-from-cran">Install from CRAN<a class="anchor" aria-label="anchor" href="#install-from-cran"></a>
</h4>
<p><a href="https://cran.r-project.org/package=volcano3D" class="external-link"><img src="https://www.r-pkg.org/badges/version/volcano3D" alt="CRAN status"></a></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"volcano3D"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="install-from-github">Install from Github<a class="anchor" aria-label="anchor" href="#install-from-github"></a>
</h4>
<p><a href="https://GitHub.com/KatrionaGoldmann/volcano3D/tags/" class="external-link"><img src="https://img.shields.io/github/tag/KatrionaGoldmann/volcano3D.svg" alt="GitHub tag"></a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"KatrionaGoldmann/volcano3D"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="load-the-package">Load the package<a class="anchor" aria-label="anchor" href="#load-the-package"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://katrionagoldmann.github.io/volcano3D/index.html" class="external-link">volcano3D</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<p>This vignette uses a subset of the 500 genes from the PEAC dataset to
explore the package functions. This can be loaded using:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_data"</span><span class="op">)</span></span></code></pre></div>
<p>Which contains:</p>
<ul>
<li><p><code>syn_example_rld</code> - the log transformed expression
data</p></li>
<li><p><code>syn_example_meta</code> which contains sample information
and divides the samples into 3 classes.</p></li>
</ul>
<p>Samples in this cohort fall into three histological ‘pathotype’
groups:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">syn_example_meta</span><span class="op">$</span><span class="va">Pathotype</span><span class="op">)</span>, col.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pathotype"</span>, <span class="st">"Count"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
Pathotype
</th>
<th style="text-align:right;">
Count
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Lymphoid
</td>
<td style="text-align:right;">
45
</td>
</tr>
<tr>
<td style="text-align:left;">
Myeloid
</td>
<td style="text-align:right;">
20
</td>
</tr>
<tr>
<td style="text-align:left;">
Fibroid
</td>
<td style="text-align:right;">
16
</td>
</tr>
</tbody>
</table>
<p>These will be used as the differential expression classes for the
three-way analysis.</p>
<div class="section level3">
<h3 id="creating-polar-coordinates">Creating Polar Coordinates<a class="anchor" aria-label="anchor" href="#creating-polar-coordinates"></a>
</h3>
<p>The function <code><a href="../reference/polar_coords.html">polar_coords()</a></code> is used to map attributes to
polar coordinates. If you have RNA-Seq count data this step can be
skipped and you can use functions <code><a href="../reference/deseq_polar.html">deseq_polar()</a></code> or
<code><a href="../reference/voom_polar.html">voom_polar()</a></code> instead (see <a href="#gene-expression-pipeline">Gene Expression pipeline</a>).</p>
<p><code>polar_coords</code> accepts raw data and performs all the
calculations needed to generate coordinates, colours etc for plotting
either a 3d volcano plot or radial 3-way plot. In brief, the function
calculates the mean of each attribute/ variable for each group and maps
the mean level per group onto polar coordinates along 3 axes in the x-y
plane. The z axis is plotted as -log<sub>10</sub>(p-value) of the group
statistical test (e.g. likelihood ratio test, one-way ANOVA or
Kruskal-Wallis test).</p>
<p>A table of p-values can be supplied by the user (see table below for
formatting requirements). If a table of p-values is absent, p-values are
automatically calculated by <code><a href="../reference/polar_coords.html">polar_coords()</a></code>. By default
one-way ANOVA is used for the group comparison and t-tests are used for
pairwise tests.</p>
<p><code><a href="../reference/polar_coords.html">polar_coords()</a></code> has the following inputs:</p>
<table class="table table" style="font-size: 11px; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
Variable
</th>
<th style="text-align:left;">
Details
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
<p>outcome</p>
(required)
</td>
<td style="text-align:left;">
Vector containing three-level factor indicating which of the three
classes each sample belongs to.
</td>
</tr>
<tr>
<td style="text-align:left;">
<p>data</p>
(required)
</td>
<td style="text-align:left;">
A dataframe or matrix containing data to be compared between the three
classes (e.g. gene expression data). Note that variables are in columns,
so gene expression data will need to be transposed. This is used to
calculate z-score and fold change, so for gene expression count data it
should be normalised such as log transformed or variance stabilised
count transformation.
</td>
</tr>
<tr>
<td style="text-align:left;">
<p>pvals</p>
(optional)
</td>
<td style="text-align:left;">
<p>the pvals matrix which contains the statistical significance of
probes or attributes between classes. This contains:</p>
<ul>
<li><p>the first column is a group test such as one-way ANOVA or
Kruskal-Wallis test.</p></li>
<li>
<p>columns 2-4 contain p-values one for each comparison in the
sequence A vs B, A vs C, B vs C, where A, B, C are the three levels in
sequence in the outcome factor. For gene expression RNA-Seq count data,
conduit functions using ‘limma voom’ or ‘DESeq’ pipelines to extract
p-values for analysis are provided in functions
<code><a href="../reference/deseq_polar.html">deseq_polar()</a></code> and <code><a href="../reference/voom_polar.html">voom_polar()</a></code>. If p-values
are not provided by the user, they can be calculated via the
<code><a href="../reference/polar_coords.html">polar_coords()</a></code> function.</p>
</li>
</ul>
</td>
</tr>
<tr>
<td style="text-align:left;">
<p>padj</p>

(optional)
</td>
<td style="text-align:left;">
Matrix containing the adjusted p-values matching the pvals matrix.
</td>
</tr>
<tr>
<td style="text-align:left;">
pcutoff
</td>
<td style="text-align:left;">
Cut-off for p-value significance
</td>
</tr>
<tr>
<td style="text-align:left;">
scheme
</td>
<td style="text-align:left;">
Vector of colours starting with non-significant attributes
</td>
</tr>
<tr>
<td style="text-align:left;">
labs
</td>
<td style="text-align:left;">
Optional character vector for labelling classes. Default
<code>NULL</code> leads to abbreviated labels based on levels in
<code>outcome</code> using <code><a href="https://rdrr.io/r/base/abbreviate.html" class="external-link">abbreviate()</a></code>. A vector of length
3 with custom abbreviated names for the outcome levels can be supplied.
Otherwise a vector length 7 is expected, of the form “ns”, “B+”, “B+C+”,
“C+”, “A+C+”, “A+”, “A+B+”, where “ns” means non-significant and A, B, C
refer to levels 1, 2, 3 in <code>outcome</code>, and must be in the
correct order.
</td>
</tr>
</tbody>
</table>
<p>This can be applied to the example data as below:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">syn_polar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/polar_coords.html">polar_coords</a></span><span class="op">(</span>outcome <span class="op">=</span> <span class="va">syn_example_meta</span><span class="op">$</span><span class="va">Pathotype</span>,</span>
<span>                          data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">syn_example_rld</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This creates a ‘volc3d’ class object for downstream plotting.</p>
</div>
<div class="section level3">
<h3 id="gene-expression-pipeline">Gene expression pipeline<a class="anchor" aria-label="anchor" href="#gene-expression-pipeline"></a>
</h3>
<p>RNA-Sequencing gene expression count data can be compared for
differentially expressed genes between 3 classes using 2 pipeline
functions to allow statistical analysis by Bioconductor packages
‘DESeq2’ and ‘limma voom’ to quickly generate a polar plotting object of
class ‘volc3d’ which can be plotted either as a 2d polar plot with 3
axes or as a 3d cylindrical plot with a 3d volcano plot.</p>
<p>Two functions <code><a href="../reference/deseq_polar.html">deseq_polar()</a></code> and
<code><a href="../reference/voom_polar.html">voom_polar()</a></code> are available. They both take RNA-Seq count
data objects as input and extract correct statistical results and then
internally call <code><a href="../reference/polar_coords.html">polar_coords()</a></code> to create a ‘volc3d’ class
object which can be plotted straightaway.</p>
<div class="section level4">
<h4 id="method-using-deseq2">Method using DESeq2<a class="anchor" aria-label="anchor" href="#method-using-deseq2"></a>
</h4>
<p>This takes 2 <code>DESeqDataSet</code> objects and converts the
results to a ‘volc3d’ class object for plotting. <code>object</code> is
an object of class ‘DESeqDataSet’ with the full design formula. Note the
function <code>DESeq</code> needs to have been previously run on this
object. <code>objectLRT</code> is an object of class ‘DESeqDataSet’ with
the reduced design formula. The function <code>DESeq</code> needs to
have been run on this object with <code>DESeq</code> argument
<code>test="LRT"</code>.</p>
<p>Note that in the DESeq2 design formula, the 3-class variable of
interest should be first.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># setup initial dataset from Tximport</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromTximport</a></span><span class="op">(</span>txi <span class="op">=</span> <span class="va">syn_txi</span>, </span>
<span>                               colData <span class="op">=</span> <span class="va">syn_metadata</span>, </span>
<span>                               design <span class="op">=</span> <span class="op">~</span> <span class="va">Pathotype</span> <span class="op">+</span> <span class="va">Batch</span> <span class="op">+</span> <span class="va">Gender</span><span class="op">)</span></span>
<span><span class="co"># initial analysis run</span></span>
<span><span class="va">dds_DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co"># likelihood ratio test on 'Pathotype'</span></span>
<span><span class="va">dds_LRT</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span>, test <span class="op">=</span> <span class="st">"LRT"</span>, reduced <span class="op">=</span> <span class="op">~</span> <span class="va">Batch</span> <span class="op">+</span> <span class="va">Gender</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># create 'volc3d' class object for plotting</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deseq_polar.html">deseq_polar</a></span><span class="op">(</span><span class="va">dds_DE</span>, <span class="va">dds_LRT</span>, <span class="st">"Pathotype"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot 3d volcano plot</span></span>
<span><span class="fu"><a href="../reference/volcano3D.html">volcano3D</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="method-using-limma-voom">Method using limma voom<a class="anchor" aria-label="anchor" href="#method-using-limma-voom"></a>
</h4>
<p>The method for limma voom is faster and takes a design formula,
metadata and raw count data. The Bioconductor packages ‘limma’ and
‘edgeR’ are used to analyse the data using the ‘voom’ method. The
results are converted to a ‘volc3d’ object ready for plotting a 3d
volcano plot or 3-way polar plot.</p>
<p>Note the design formula must be of the form
<code>~ 0 + outcome + ...</code>. The 3-class outcome variable must be
the first variable after the ‘0’, and this variable must be a factor
with exactly 3 levels.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">syn_tpm</span> <span class="op">&lt;-</span> <span class="va">syn_txi</span><span class="op">$</span><span class="va">counts</span>  <span class="co"># raw counts</span></span>
<span></span>
<span><span class="va">resl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/voom_polar.html">voom_polar</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">Pathotype</span> <span class="op">+</span> <span class="va">Batch</span> <span class="op">+</span> <span class="va">Gender</span>, <span class="va">syn_metadata</span>, <span class="va">syn_tpm</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/volcano3D.html">volcano3D</a></span><span class="op">(</span><span class="va">resl</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="radial-plots">Radial Plots<a class="anchor" aria-label="anchor" href="#radial-plots"></a>
</h3>
<p>The differential expression can now be visualised on an interactive
radial plot using <code>radial_plotly</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/radial_plotly.html">radial_plotly</a></span><span class="op">(</span><span class="va">syn_polar</span><span class="op">)</span> </span></code></pre></div>
<p>Unfortunately CRAN does not support interactive plotly in the
vignette, but these can be viewed on the <a href="https://katrionagoldmann.github.io/volcano3D/articles/Extended_Vignette.html" class="external-link">extended
vignette</a>. When interactive, it is possible to identify genes for
future interrogation by hovering over certain markers.</p>
<p><code>radial_plotly</code> produces an SVG based plotly object by
default. With 10,000s of points SVG can be slow, so for large number of
points we recommend switching to webGL by piping the plotly object to
<code><a href="https://rdrr.io/pkg/plotly/man/toWebGL.html" class="external-link">toWebGL()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/radial_plotly.html">radial_plotly</a></span><span class="op">(</span><span class="va">syn_polar</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/plotly/man/toWebGL.html" class="external-link">toWebGL</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>A very similar looking static ggplot image can be created using
<code>radial_ggplot</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/radial_ggplot.html">radial_ggplot</a></span><span class="op">(</span><span class="va">syn_polar</span>,</span>
<span>              marker_size <span class="op">=</span> <span class="fl">2.3</span>,</span>
<span>              legend_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Vignette_files/figure-html/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="boxplots">Boxplots<a class="anchor" aria-label="anchor" href="#boxplots"></a>
</h3>
<p>Any one specific variable (such as a gene) can be interrogated using
a boxplot to investigate differences between groups:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/boxplot_trio.html">boxplot_trio</a></span><span class="op">(</span><span class="va">syn_polar</span>,</span>
<span>                      value <span class="op">=</span> <span class="st">"COBL"</span>,</span>
<span>                      text_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>                      test <span class="op">=</span> <span class="st">"polar_padj"</span>,</span>
<span>                      my_comparisons<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Lymphoid"</span>, <span class="st">"Myeloid"</span><span class="op">)</span>,</span>
<span>                                          <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Lymphoid"</span>, <span class="st">"Fibroid"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/boxplot_trio.html">boxplot_trio</a></span><span class="op">(</span><span class="va">syn_polar</span>,</span>
<span>                      value <span class="op">=</span> <span class="st">"COBL"</span>,</span>
<span>                      box_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"violet"</span>, <span class="st">"gold2"</span><span class="op">)</span>,</span>
<span>                      levels_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Lymphoid"</span>, <span class="st">"Fibroid"</span><span class="op">)</span>,</span>
<span>                      text_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>                      test <span class="op">=</span> <span class="st">"polar_padj"</span></span>
<span>                      <span class="op">)</span></span>
<span></span>
<span><span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/boxplot_trio.html">boxplot_trio</a></span><span class="op">(</span><span class="va">syn_polar</span>,</span>
<span>                      value <span class="op">=</span> <span class="st">"TREX2"</span>,</span>
<span>                      text_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>                      stat_size<span class="op">=</span><span class="fl">2.5</span>,</span>
<span>                      test <span class="op">=</span> <span class="st">"polar_multi_padj"</span>,</span>
<span>                      levels_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Lymphoid"</span>, <span class="st">"Myeloid"</span>, <span class="st">"Fibroid"</span><span class="op">)</span>,</span>
<span>                      box_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"green3"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span><span class="va">plot1</span>, <span class="va">plot2</span>, <span class="va">plot3</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="Vignette_files/figure-html/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="d-volcano-plot">3D Volcano Plot<a class="anchor" aria-label="anchor" href="#d-volcano-plot"></a>
</h3>
<p>Lastly, the 3D volcano plot can be used to project 3-way differential
comparisons such as differential gene expression between 3 classes onto
cylindrical polar coordinates.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/volcano3D.html">volcano3D</a></span><span class="op">(</span><span class="va">syn_polar</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="volcano3D.png" width="80%" style="border: 0;"></p>
<p>A fully interactive demonstration of this plot can be viewed at: <a href="https://peac.hpc.qmul.ac.uk/" class="external-link uri">https://peac.hpc.qmul.ac.uk/</a></p>
</div>
<div class="section level3">
<h3 id="spinning-animation">Spinning Animation<a class="anchor" aria-label="anchor" href="#spinning-animation"></a>
</h3>
<p>It is also possible to animate spinning/rotation of the plot in 3d
using the <code>add_animation</code> function which adds a custom plotly
modeBar button:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/add_animation.html">add_animation</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p>A working demo of this function can be viewed at <a href="https://peac.hpc.qmul.ac.uk/" class="external-link uri">https://peac.hpc.qmul.ac.uk/</a> by clicking on the ‘play’
button in the plotly modeBar to spin the plot.</p>
</div>
<div class="section level3">
<h3 id="saving-plotly-plots">Saving Plotly Plots<a class="anchor" aria-label="anchor" href="#saving-plotly-plots"></a>
</h3>
<p><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></p>
<div class="section level4">
<h4 id="static-images">Static Images<a class="anchor" aria-label="anchor" href="#static-images"></a>
</h4>
<p>There are a few ways to save plotly plots as static images. Firstly
plotly offers a download button (
<i class="fa fa-camera" aria-hidden="true"></i> ) in the figure mode bar
(appears top right). By default this saves images as png, however it is
possible to convert to svg, jpeg or webp using:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">plotly</span><span class="fu">::</span><span class="fu">config</span><span class="op">(</span>toImageButtonOptions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>format <span class="op">=</span> <span class="st">"svg"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="interactive-html">Interactive HTML<a class="anchor" aria-label="anchor" href="#interactive-html"></a>
</h4>
<p>The full plotly objects can be saved to HTML by converting them to
widgets and saving with the htmlwidgets package:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">htmlwidgets</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/htmlwidgets/man/saveWidget.html" class="external-link">saveWidget</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/plotly/man/as_widget.html" class="external-link">as_widget</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span>, <span class="st">"volcano3D.html"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>volcano3D was developed by the bioinformatics team at the <a href="https://www.qmul.ac.uk/whri/emr/" class="external-link">Experimental Medicine &amp;
Rheumatology department</a> and <a href="https://www.qmul.ac.uk/c4tb/" class="external-link">Centre for Translational
Bioinformatics</a> at Queen Mary University London.</p>
<p>If you use this package please cite as:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"volcano3D"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## To cite package 'volcano3D' in publications use:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   Goldmann K, Lewis M (2020). _volcano3D: 3D Volcano Plots and Polar</span></span>
<span><span class="co">##   Plots for Three-Class Data_.</span></span>
<span><span class="co">##   https://katrionagoldmann.github.io/volcano3D/index.html,</span></span>
<span><span class="co">##   https://github.com/KatrionaGoldmann/volcano3D.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##   @Manual{,</span></span>
<span><span class="co">##     title = {volcano3D: 3D Volcano Plots and Polar Plots for Three-Class Data},</span></span>
<span><span class="co">##     author = {Katriona Goldmann and Myles Lewis},</span></span>
<span><span class="co">##     year = {2020},</span></span>
<span><span class="co">##     note = {https://katrionagoldmann.github.io/volcano3D/index.html, https://github.com/KatrionaGoldmann/volcano3D},</span></span>
<span><span class="co">##   }</span></span></code></pre>
<p>or using:</p>
<blockquote>
<p>Lewis, Myles J., et al. <em>Molecular portraits of early rheumatoid
arthritis identify clinical and treatment response phenotypes</em>. Cell
reports 28.9 (2019): 2455-2470.</p>
</blockquote>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Katriona Goldmann, Myles Lewis.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
