<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert RNA-Seq count data to a volcano3d object using 'limma voom' — voom_polar • volcano3D</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Convert RNA-Seq count data to a volcano3d object using 'limma voom' — voom_polar" />
<meta property="og:description" content="This function is used instead of polar_coords if you have raw
RNA-Seq count data. The function takes a design formula, metadata and raw
RNA-Seq count data and uses 'limma voom' to analyse the data. The results are
converted to a 'volc3d' object ready for plotting a 3d volcano plot or polar
plot." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-158679490-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158679490-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">volcano3D</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fas fa-book"></span>
     
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Extended_Vignette.html">Volcano3D Vignette</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Additional Tutorials</li>
    <li>
      <a href="../articles/Vignette_2x3.html">volcano3D: 2x3-way analysis</a>
    </li>
    <li>
      <a href="../articles/shiny_builder.html">Using Volcano3D to create Shiny Apps</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">
    <span class="fas fa-newspaper"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="far fa-file-code"></span>
     
    Reference Manual
  </a>
</li>
<li>
  <a href="https://github.com/KatrionaGoldmann/volcano3D">
    <span class="fab fa-github"></span>
     
    github
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert RNA-Seq count data to a volcano3d object using 'limma voom'</h1>
    <small class="dont-index">Source: <a href='https://github.com/KatrionaGoldmann/volcano3D/blob/master/R/voom_polar.R'><code>R/voom_polar.R</code></a></small>
    <div class="hidden name"><code>voom_polar.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function is used instead of <code><a href='polar_coords.html'>polar_coords</a></code> if you have raw
RNA-Seq count data. The function takes a design formula, metadata and raw
RNA-Seq count data and uses 'limma voom' to analyse the data. The results are
converted to a 'volc3d' object ready for plotting a 3d volcano plot or polar
plot.</p>
    </div>

    <pre class="usage"><span><span class='fu'>voom_polar</span><span class='op'>(</span></span>
<span>  <span class='va'>formula</span>,</span>
<span>  <span class='va'>metadata</span>,</span>
<span>  <span class='va'>counts</span>,</span>
<span>  pcutoff <span class='op'>=</span> <span class='fl'>0.05</span>,</span>
<span>  padj.method <span class='op'>=</span> <span class='st'>"BH"</span>,</span>
<span>  filter_pairwise <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>  <span class='va'>...</span></span>
<span><span class='op'>)</span></span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>formula</th>
      <td><p>Design formula which must be of the form <code>~ 0 + outcome + ...</code>. The 3-way outcome variable must be the first variable after the '0',
and this variable must be a factor with exactly 3 levels.</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p>Matrix or dataframe containing metadata as referenced by
<code>formula</code></p></td>
    </tr>
    <tr>
      <th>counts</th>
      <td><p>Matrix containing raw gene expression count data</p></td>
    </tr>
    <tr>
      <th>pcutoff</th>
      <td><p>Cut-off for p-value significance</p></td>
    </tr>
    <tr>
      <th>padj.method</th>
      <td><p>Can be any method available in <code>p.adjust</code> or <code>"qvalue"</code>.
The option "none" is a pass-through.</p></td>
    </tr>
    <tr>
      <th>filter_pairwise</th>
      <td><p>Logical whether adjusted p-value pairwise statistical
tests are only conducted on genes which reach significant adjusted p-value
cut-off on the group likelihood ratio test</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Optional arguments passed to <code><a href='polar_coords.html'>polar_coords</a></code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Calls <code><a href='polar_coords.html'>polar_coords</a></code> to return an S4 'volc3d' object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Statistical results for the group and pairwise comparisons are calculated
using the 'limma voom' pipeline and the results passed to
<code><a href='polar_coords.html'>polar_coords</a></code> to generate a 'volc3d' object ready for plotting a
3d volcano plot or polar plot.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='polar_coords.html'>polar_coords</a></code>, <code><a href='deseq_polar.html'>deseq_polar</a></code>,
<code><a href='https://rdrr.io/pkg/limma/man/voom.html'>voom</a></code> in the limma package</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'><span><span class='kw'>if</span> <span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/ns-load.html'>requireNamespace</a></span><span class='op'>(</span><span class='st'>"limma"</span>, quietly <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>&amp;</span> </span></span>
<span class='r-in'><span>    <span class='fu'><a href='https://rdrr.io/r/base/ns-load.html'>requireNamespace</a></span><span class='op'>(</span><span class='st'>"edgeR"</span>, quietly <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span></span></span>
<span class='r-in'><span>  <span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/limma'>limma</a></span><span class='op'>)</span></span></span>
<span class='r-in'><span>  <span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/edgeR'>edgeR</a></span><span class='op'>)</span></span></span>
<span class='r-in'><span>  </span></span>
<span class='r-in'><span>  <span class='va'>counts</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span><span class='op'>(</span>n<span class='op'>=</span><span class='fl'>1500</span>, mu<span class='op'>=</span><span class='fl'>100</span>, size<span class='op'>=</span><span class='fl'>1</span><span class='op'>/</span><span class='fl'>0.5</span><span class='op'>)</span>, ncol<span class='op'>=</span><span class='fl'>15</span><span class='op'>)</span></span></span>
<span class='r-in'><span>  <span class='va'>cond</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>, each<span class='op'>=</span><span class='fl'>5</span><span class='op'>)</span>, labels <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>'A'</span>, <span class='st'>'B'</span>, <span class='st'>'C'</span><span class='op'>)</span><span class='op'>)</span></span></span>
<span class='r-in'><span>  <span class='va'>cond</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span><span class='va'>cond</span><span class='op'>)</span></span></span>
<span class='r-in'><span>  </span></span>
<span class='r-in'><span>  <span class='va'>polar</span> <span class='op'>&lt;-</span> <span class='fu'>voom_polar</span><span class='op'>(</span><span class='op'>~</span><span class='fl'>0</span> <span class='op'>+</span> <span class='va'>cond</span>, <span class='va'>cond</span>, <span class='va'>counts</span><span class='op'>)</span></span></span>
<span class='r-in'><span>  </span></span>
<span class='r-in'><span>  <span class='fu'><a href='volcano3D.html'>volcano3D</a></span><span class='op'>(</span><span class='va'>polar</span><span class='op'>)</span></span></span>
<span class='r-in'><span>  <span class='fu'><a href='radial_ggplot.html'>radial_ggplot</a></span><span class='op'>(</span><span class='va'>polar</span><span class='op'>)</span></span></span>
<span class='r-in'><span><span class='op'>}</span></span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Attaching package: ‘limma’</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:DESeq2’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     plotMA</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> The following object is masked from ‘package:BiocGenerics’:</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span>     plotMA</span>
<span class='r-plt img'><img src='voom_polar-1.png' alt='' width='700' height='433' /></span>
<span class='r-in'><span></span></span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css">

<style>
  #left {
    float:left;
  }
  #right {
    width:100%;
    position:relative;
    text-align:right;
    align:right;
    float:right;
  }
  *.icon-grey {color: lightgrey}
  j {
    text-align:center;
    vertical-align:middle;
    position: relative;
    padding-bottom: 12px;
  }
  .smadge:after{
    content:"EMR";
    position: absolute;
    background: lightgrey;
    height:1.2rem;
    top:1rem;
    right:-1rem;
    width:2.3rem;
    text-align: center;
    line-height: 1rem;;
    font-size: 1rem;
    color: #cc0099;
    border-radius: 10%;
  }
  .bladge:after{
    content:"C4TB";
    position: absolute;
    background: lightgrey;
    height:1.2rem;
    top:1rem;
    right:-1rem;
    width:2.7rem;
    text-align: center;
    line-height: 1rem;;
    font-size: 1rem;
    color: #cc0099;
    border-radius: 10%;
  }
</style>

<div id="left">
Developed by <a href="https://www.turing.ac.uk/people/enrichment-students/katriona-goldmann">Katriona Goldmann</a> and <a href="https://www.qmul.ac.uk/whri/people/academic-staff/items/lewismyles.html">Myles Lewis</a>
</div>

<div id="right">
<a href="https://www.qmul.ac.uk/whri/emr/"><j class="fa fa-university fa-2x  icon-grey smadge" aria-hidden="false" style="color:#cc0099"></j></a>
<span style="margin-left:1rem"></span>

<a href="https://www.qmul.ac.uk/c4tb/"><j class="fa fa-university fa-2x  icon-grey bladge" aria-hidden="false" style="color:#cc0099"></j></a>
<span style="margin-left:1rem"></span>

<a href="https://towardsdatascience.com/interactive-differential-expression-analysis-with-volcano3d-cf6c37b5aa6a"><i class="fab fa-medium-m fa-2x" style="color:#cc0099"></i></a>
<span style="margin-left:1rem"></span>
<a href="mailto:k.goldmann@qmul.ac.uk" class="fas fa-envelope fa-2x" style="color:#cc0099">
</a><span style="margin-left:1rem"></span>

<a href="https://github.com/KatrionaGoldmann"><i class="fab fa-github fa-2x" style="color:#cc0099"></i></a>
<span style="margin-left:1rem"></span>

<a href="https://twitter.com/_katriona"><i class="fab fa-twitter fa-2x" style="color:#cc0099"></i></a>
</div>

      </footer>
   </div>

  


  </body>
</html>


